server/src/main/scala/services/graphstages/WorkloadGraphStage.scala

```diff
@@ -0,0 +1,157 @@
+      initialFlights.map(fws => {
+        val uniqueFlightId = fws.apiFlight.uniqueId
+        val flightWorkload = WorkloadCalculator.flightToFlightSplitMinutes(fws, airportConfig.defaultProcessingTimes.head._2, natProcTimes, useNationalityBasedProcessingTimes)
+
+        (uniqueFlightId, flightWorkload)
+      }).toMap
+    }
+
+    setHandler(inFlightsWithSplits, new InHandler {
+      override def onPush(): Unit = {
+        val incomingFlights = grab(inFlightsWithSplits)
+        log.info(s"Received ${incomingFlights.flights.size} arrivals $incomingFlights")
+        /*
+        1) map over each incoming flights to produce new workload
+        2) update flightId -> workload map
+        3) merge all workloads to Set[LoadMinute]
+        4) calc new/old LoadMinute diff and push
+         */
+
+        val updatedWorkloads: Map[Int, Set[FlightSplitMinute]] = mergeWorkloadByFlightId(incomingFlights, workloadByFlightId)
+        log.info(s"updatedWorkloads: $updatedWorkloads")
+        workloadByFlightId = purgeExpired(updatedWorkloads, (fsms: Set[FlightSplitMinute]) => fsms.map(_.minute).min, now, expireAfterMillis)
+        val updatedLoads: Set[LoadMinute] = flightSplitMinutesToQueueLoadMinutes(updatedWorkloads)
+
+        val diff = loadDiff(updatedLoads, loadMinutes)
+        loadMinutes = purgeExpired(updatedLoads, (lm: LoadMinute) => lm.minute, now, expireAfterMillis)
+
+        loadsToPush = purgeExpired(mergeLoadMinutes(diff, loadsToPush), (lm: LoadMinute) => lm.minute, now, expireAfterMillis)
+        log.info(s"Now have ${loadsToPush.size} load minutes to push")
+
+        pushStateIfReady()
+
+        pullFlights()
+      }
+    })
+
+    def mergeLoadMinutes(updatedLoads: Set[LoadMinute], existingLoads: Map[Int, LoadMinute]): Map[Int, LoadMinute] = {
+      updatedLoads.foldLeft(existingLoads) {
+        case (soFar, newLoadMinute) => soFar.updated(newLoadMinute.uniqueId, newLoadMinute)
+      }
+    }
+
+    def loadDiff(updatedLoads: Set[LoadMinute], existingLoads: Set[LoadMinute]): Set[LoadMinute] = {
+      val updates = updatedLoads -- existingLoads
+      val removeIds = (existingLoads.map(_.uniqueId) -- updatedLoads.map(_.uniqueId))
+      val removes = existingLoads.filter(l => removeIds.contains(l.uniqueId)).map(_.copy(paxLoad = 0, workLoad = 0))
+      val diff = updates ++ removes
+      log.info(s"${diff.size} updated load minutes")
+
+      diff
+    }
+
+    def mergeWorkloadByFlightId(incomingFlights: FlightsWithSplits, existingLoads: Map[Int, Set[FlightSplitMinute]]): Map[Int, Set[FlightSplitMinute]] = {
+      incomingFlights.flights.foldLeft(existingLoads) {
+        case (soFar, fws) =>
+          val uniqueFlightId = fws.apiFlight.uniqueId
+          val flightWorkload = WorkloadCalculator.flightToFlightSplitMinutes(fws, airportConfig.defaultProcessingTimes.head._2, natProcTimes, useNationalityBasedProcessingTimes)
```

Comment:
what is head._2 here? I think defaultProcessingTimes is a Map, so this seems a bit weird.

Reply:
yeah legacy code from the existing crunch stuff that i didn't clean up...